lxd/patches.go

```diff
@@ -854,40 +870,55 @@ func upgradeFromStorageTypeZfs(name string, d *Daemon, defaultPoolName string, d
 
 	if len(cRegular) > 0 {
 		containersSubvolumePath := getContainerMountPoint(poolName, "")
-		err := os.MkdirAll(containersSubvolumePath, 0711)
-		if err != nil {
-			return err
+		if !shared.PathExists(containersSubvolumePath) {
+			err := os.MkdirAll(containersSubvolumePath, 0711)
+			if err != nil {
+				shared.LogWarnf("Failed to create path: %s.", containersSubvolumePath)
+			}
 		}
 	}
 
+	failedUpgradeEntities := []string{}
 	for _, ct := range cRegular {
-
-		// Insert storage volumes for containers into the database.
-		_, err := dbStoragePoolVolumeCreate(d.db, ct, storagePoolVolumeTypeContainer, poolID, volumeConfig)
-		if err != nil {
-			shared.LogWarnf("Could not insert a storage volume for container \"%s\".", ct)
-			continue
+		_, err := dbStoragePoolVolumeGetTypeID(d.db, ct, storagePoolVolumeTypeContainer, poolID)
+		if err == nil {
+			shared.LogWarnf("Storage volumes database already contains an entry for the container.")
+		} else if err == NoSuchObjectError {
+			// Insert storage volumes for containers into the database.
+			_, err := dbStoragePoolVolumeCreate(d.db, ct, storagePoolVolumeTypeContainer, poolID, volumeConfig)
+			if err != nil {
+				// In case we're on a LXD instance where a a partial
+				// upgrade took place, we should still proceed.
+				shared.LogErrorf("Could not insert a storage volume for container \"%s\".", ct)
+				return err
+			}
+		} else {
+			shared.LogErrorf("Failed to query database: %s", err)
+			return err
 		}
 
 		// Unmount the container zfs doesn't really seem to care if we
 		// do this.
 		ctDataset := fmt.Sprintf("%s/containers/%s", defaultPoolName, ct)
 		oldContainerMntPoint := shared.VarPath("containers", ct)
 		if shared.IsMountPoint(oldContainerMntPoint) {
-			output, err := tryExec("zfs", "unmount", "-f", ctDataset)
+			_, err := tryExec("zfs", "unmount", "-f", ctDataset)
 			if err != nil {
-				return fmt.Errorf("Failed to unmount ZFS filesystem: %s", output)
+				shared.LogWarnf("Failed to unmount ZFS filesystem via zfs unmount. Trying lazy umount (MNT_DETACH)...")
+				err := tryUnmount(oldContainerMntPoint, syscall.MNT_DETACH)
+				if err != nil {
+					failedUpgradeEntities = append(failedUpgradeEntities, fmt.Sprintf("containers/%s: Failed to umount zfs filesystem.", ct))
+					continue
+				}
 			}
 		}
 
-		err = os.Remove(oldContainerMntPoint)
-		if err != nil {
-			return err
+		if shared.PathExists(oldContainerMntPoint) {
```

Comment:
Those PathExists() checks can be dropped if you're not going to check the return value of os.Remove anyway.